parameter/criterion/src/Criterion.cpp

```diff
@@ -0,0 +1,296 @@
+ * are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation and/or
+ * other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "criterion/Criterion.h"
+#include <Utility.h>
+#include <log/Logger.h>
+
+#include <stdexcept>
+#include <sstream>
+#include <cassert>
+
+namespace core
+{
+namespace selection
+{
+namespace criterion
+{
+
+Criterion::Criterion(const std::string& name, const Values& values, core::log::Logger& logger)
+    : Criterion(name, logger, values, State{values.begin()->second}, {})
+{
+    if (mValues.size() < 2) {
+        throw InvalidCriterionError("Not enough values were provided for exclusive criterion '" +
+                                    mName  + "' which needs at least 2 values");
+    }
+}
+
+Criterion::Criterion(const std::string& name,
+                     core::log::Logger& logger,
+                     const Values& values,
+                     const State&& defaultState,
+                     const MatchMethods& derivedMatchMethods)
+    : mValues(values),
+      mMatchMethods(gatherMap<MatchMethods>(MatchMethods{
+                                {"Is", [&](const State& state){ return mState == state; }},
+                                {"IsNot", [&](const State& state){ return mState != state; }}},
+                                derivedMatchMethods)),
+      mState(), _uiNbModifications(0), _logger(logger), mName(name), mDefaultState(defaultState)
+{
+    setDefaultState();
+}
+
+bool Criterion::hasBeenModified() const
+{
+    return _uiNbModifications != 0;
+}
+
+void Criterion::resetModifiedStatus()
+{
+    _uiNbModifications = 0;
+}
+
+bool Criterion::setDefaultState()
```

Comment:
see below. IMO, the code for this method should be:

    return setState(mDefaultState);